packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+     * property's value type, `false` otherwise
+     */
+    function indexSignatureHasMatchingValue(
+      remainingProperties: RemainingProperties,
+      indexSignature: TSESTree.TSIndexSignature,
+    ): boolean {
+      // `{ used }: { [i: `_${string}`] }`
+      if (!indexSignature.typeAnnotation) {
+        return true;
+      }
+
+      const indexType = services.getTypeAtLocation(
+        indexSignature.typeAnnotation.typeAnnotation,
+      );
+
+      for (const destructure of remainingProperties.values()) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.value,
+        );
+
+        if (isTypeMatchingIndexSignatureType(destructure.type, indexType)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * Compares an index-signature value-type with another type. If `isNoUncheckedIndexedAccess`
+     * is enabled, will filter out `undefined` from the index-signature's type.
+     */
+    function isTypeMatchingIndexSignatureType(
+      indexSignatureType: ts.Type,
+      comparedType: ts.Type,
+    ): boolean {
+      if (isNoUncheckedIndexedAccess) {
+        const indexSignatureParts = tsutils
+          .unionTypeParts(indexSignatureType)
+          .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+        const comparedTypeParts = tsutils.unionTypeParts(comparedType);
+
+        if (indexSignatureParts.length !== comparedTypeParts.length) {
+          return false;
+        }
+
+        const indexSignaturePartsSet = new Set(indexSignatureParts);
+
+        return comparedTypeParts.every(part =>
+          indexSignaturePartsSet.has(part),
+        );
+      }
+
+      return indexSignatureType === comparedType;
+    }
+
+    function reportOnMember(
+      member: TSESTree.Node,
+      data: { key: string; type: string },
+    ): void {
+      context.report({
+        node: member,
+        messageId: 'partialDestructuring',
+        data,
+        suggest: [
+          {
+            messageId: 'removeUnusedKey',
+            data,
+            fix: fixer => {
+              const nextToken = context.sourceCode.getTokenAfter(member);
+
+              if (nextToken?.value === ',') {
+                return fixer.removeRange([member.range[0], nextToken.range[1]]);
+              }
+
+              return fixer.remove(member);
+            },
+          },
+        ],
+      });
+    }
+
+    return {
+      "ArrayPattern[typeAnnotation.typeAnnotation.type='TSTupleType']"(
+        node: TSESTree.ArrayPattern & {
+          typeAnnotation: TSESTree.TSTypeAnnotation;
+        },
+      ): void {
+        checkParam(node, node.typeAnnotation.typeAnnotation);
+      },
+      "ObjectPattern[typeAnnotation.typeAnnotation.type='TSTypeLiteral']"(
+        node: TSESTree.ObjectPattern & {
+          typeAnnotation: TSESTree.TSTypeAnnotation;
+        },
+      ): void {
+        checkParam(node, node.typeAnnotation.typeAnnotation);
```

Comment:
[Refactor] This one too:

```suggestion
        checkObject(node, node.typeAnnotation.typeAnnotation);
```